segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java

```diff
@@ -652,14 +654,22 @@ private void triggerCacheTailIndex(DirectSegmentAccess segment, long tailCaching
         // Begin a loop which will end when either we've reached the target offset (defined in the RecoveryTask) or when
         // the recovery task itself has been cancelled (i.e., segment evicted, shutting down, etc.).
         Map<UUID, Long> tailCachePreIndexVersionTracker = new ConcurrentHashMap<>();
+        TailUpdates updates = new TailUpdates();
         Futures.loop(
                 () -> !task.task.isDone() && preIndexOffset.get() < task.triggerIndexOffset,
                 () -> {
                     int maxLength = (int) Math.min(this.config.getMaxTailCachePreIndexBatchLength(), task.triggerIndexOffset - preIndexOffset.get());
-                    return preIndexBatch(segment, preIndexOffset.get(), maxLength, tailCachePreIndexVersionTracker, lastIndexedOffset)
+                    return preIndexBatch(segment, preIndexOffset.get(), maxLength, tailCachePreIndexVersionTracker, lastIndexedOffset, updates)
                             .thenAccept(preIndexOffset::set);
                 },
                 this.executor)
+                .thenAcceptAsync(v -> {
+                    // Atomically update all the entries in cache after performing tail-caching process.
```

Comment:
:+1:I liked that we are now adding all keys to cache atomically at the end of processing all batches. Much safer approach.